### PR TITLE
Add API client for Azure custom vnet

### PR DESCRIPTION
**Pull request by cpuguy83**

### CODE REVIEW

File: providers/azure/aci.go [hunk 1]
```diff
@@ -192,6 +205,19 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
...
+		p.subnetName = subnetName
+	}
+	if subnetCIDR := os.Getenv("ACI_SUBNET_CIDR"); subnetCIDR != "" {
+		if _, _, err := net.ParseCIDR(subnetCIDR); err != nil {
+			return nil, fmt.Errorf("error parsing provided subnet range: %v", err)
+		}
+		p.subnetCIDR = subnetCIDR
+	}
+
+	// TODO(cpuguy83): run setupSubnet and wire this up to the container group
```
Comment (robbiezhang) [inline providers/azure/aci.go]: setupSubnet is not called.

File: providers/azure/client/network/subnet.go [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+		return Subnet{}, errors.Wrap(err, "sending subnet get request failed")
+	}
+	defer resp.Body.Close()
+
+	// 200 (OK) is a success response.
+	if err := api.CheckResponse(resp); err != nil {
+		return Subnet{}, err
+	}
+
+	var subnet Subnet
```
Comment (robbiezhang) [inline providers/azure/client/network/subnet.go]: handle 404, and return nil

File: providers/azure/aci.go [hunk 3]
```diff
@@ -200,6 +226,76 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
...
+		createSubnet = true
+	}
+
+	if createSubnet {
+		subnet := network.Subnet{
+			Name: p.subnetName,
+			Properties: network.SubnetProperties{
+				AddressPrefix: p.subnetCIDR,
+				Delegations: []network.Delegation{
+					{Name: "aciDelegation", Properties: network.DelegationProperties{
```
Comment (robbiezhang) [inline providers/azure/aci.go]: new lines

File: providers/azure/aci.go (hunk 1 above)
Comment (robbiezhang) [reply providers/azure/aci.go]: the comment is no longer needed.

File: providers/azure/aci.go [hunk 4]
```diff
@@ -200,13 +230,89 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
...
 
+func (p *ACIProvider) setupNetworkProfile(auth *client.Authentication) error {
+	c, err := network.NewClient(auth)
+	if err != nil {
+		return fmt.Errorf("error creating azure networking client: %v", err)
+	}
+
+	var createSubnet bool
+
+	subnet, _ := c.GetSubnet(p.resourceGroup, p.vnetName, p.subnetName)
```
Comment (robbiezhang) [inline providers/azure/aci.go]: handle the error case.
what's if the GetSubnet return 404?

File: providers/azure/aci.go [hunk 5]
```diff
@@ -200,13 +230,89 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
...
+	if subnet.ID == "" {
+		createSubnet = true
+	} else {
+		if p.subnetCIDR != subnet.Properties.AddressPrefix {
+			return fmt.Errorf("found existing subnet with different CIDR")
+		}
+		createSubnet = true
```
Comment (robbiezhang) [inline providers/azure/aci.go]: check the delegations here: if the subnet is already delegated to ACI, we don't need to update it.

File: providers/azure/aci.go [hunk 6]
```diff
@@ -200,13 +230,89 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
...
+	}
+
+	if createSubnet {
+		subnet := network.Subnet{
```
Comment (robbiezhang) [inline providers/azure/aci.go]: 2 issues:
1. don't re-declare the subnet here. CreateOrUpdate in L265 uses the subnet defined outside the scope.
2. update should based on subnet get from L241, but update the delegations field

File: providers/azure/aci.go [hunk 7]
```diff
@@ -200,13 +230,89 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
...
+				},
+			},
+		}
+
+		if err = c.CreateOrUpdateSubnet(p.resourceGroup, p.vnetName, subnet); err != nil {
+			return fmt.Errorf("error creating subnet: %v", err)
+		}
+	}
+
+	if profile, _ := c.GetProfile(p.resourceGroup, p.nodeName); profile.ID != "" {
```
Comment (robbiezhang) [inline providers/azure/aci.go]: handles the error

File: providers/azure/client/network/profile.go [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+	Subnet ID
+}
+
+// ID is a generic struct for objets with an ID
+type ID struct {
+	ID string
+}
+
+// GetProfile gets the network profile with the provided name
+func (c *Client) GetProfile(resourceGroup, name string) (Profile, error) {
```
Comment (robbiezhang) [inline providers/azure/client/network/profile.go]: shall we return pointer to avoid struct copy?

File: providers/azure/client/network/profile.go [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+	}
+	var p Profile
+	if err := json.NewDecoder(resp.Body).Decode(&p); err != nil {
+		return Profile{}, errors.Wrap(err, "decoding get network profile response body failed")
+	}
+	return p, nil
+}
+
+// CreateOrUpdateProfile creates or updates an Azure network profile
+func (c *Client) CreateOrUpdateProfile(resourceGroup string, p Profile) error {
```
Comment (robbiezhang) [inline providers/azure/client/network/profile.go]: shall we pass in pointer? (*Profile)

File: providers/azure/client/network/subnet.go [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+// DelegationProperties stores the properties for a delegation
+type DelegationProperties struct {
+	ServiceName string
+	Actions     []string
+}
+
+// GetSubnet gets the subnet from the specified resourcegroup/vnet
+func (c *Client) GetSubnet(resourceGroup, vnet, name string) (Subnet, error) {
```
Comment (robbiezhang) [inline providers/azure/client/network/subnet.go]: return pointer (*Subnet)?

File: providers/azure/client/network/subnet.go [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+
+	var subnet Subnet
+	if err := json.NewDecoder(resp.Body).Decode(&subnet); err != nil {
+		return subnet, err
+	}
+	return subnet, nil
+}
+
+// CreatreOrUpdateSubnet creates a new or updates an existing subnet in the defined resourcegroup/vnet
+func (c *Client) CreateOrUpdateSubnet(resourceGroup, vnet string, subnet Subnet) error {
```
Comment (robbiezhang) [inline providers/azure/client/network/subnet.go]: pass in *Subnet

**Comment by cpuguy83** [comment]
Updated

File: providers/vic/vic_provider.go [hunk 12]
```diff
@@ -49,7 +49,7 @@ type VicProvider struct {
 	client         *client.PortLayer
 	imageStore     proxy.ImageStore
 	isolationProxy proxy.IsolationProxy
-	systemProxy    vicproxy.VicSystemProxy
+	systemProxy    *vicproxy.VicSystemProxy
```
[the remark this answers is not in the document]
Comment (cpuguy83) [reply providers/vic/vic_provider.go]: "dep ensure" to  updated the Vic repo which required this change.

File: providers/azure/client/network/profile.go [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+	}
+
+	// Create the url.
+	uri := api.ResolveRelative(baseURI, profilePath)
+	uri += "?" + url.Values(urlParams).Encode()
+
+	// Create the request.
+	b, err := json.Marshal(p)
+	if err != nil {
+		return errors.Wrap(err, "marshallig networking profile failed")
```
Comment (robbiezhang) [inline providers/azure/client/network/profile.go]: typo: marshalling network profile

File: providers/azure/aci.go [hunk 14]
```diff
@@ -192,6 +206,21 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
 		p.pods = podsQuota
 	}
 
+	p.subnetName = "virtual-kubelet-aci-connector"
```
Comment (robbiezhang) [inline providers/azure/aci.go]: remove the default value. If it's not set, the customer doesn't want to use VNet. you can skip setupNetworkProfile, then

File: providers/azure/aci.go [hunk 15]
```diff
@@ -200,13 +233,111 @@ func NewACIProvider(config string, rm *manager.ResourceManager, nodeName, operat
...
+}
+
 // CreatePod accepts a Pod definition and creates
 // an ACI deployment
 func (p *ACIProvider) CreatePod(pod *v1.Pod) error {
 	var containerGroup aci.ContainerGroup
 	containerGroup.Location = p.region
 	containerGroup.RestartPolicy = aci.ContainerGroupRestartPolicy(pod.Spec.RestartPolicy)
 	containerGroup.ContainerGroupProperties.OsType = aci.OperatingSystemTypes(p.OperatingSystem())
+	containerGroup.NetworkProfile.ID = p.networkProfile
```
Comment (robbiezhang) [inline providers/azure/aci.go]: NRE from NetworkProfile.ID

**Pull request merged**